File: string/string-1/49.cpp
// with sorting
// Tc = O(k.logk) -> sorting + n (iterating in a given string)
// Tc = O(n*k*logk)
//S.C : O(n*k)
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int n = strs.size();
        vector<vector<string>> result;
        unordered_map<string, vector<string>> mp;
        //n
        for (int i = 0; i < n; i++) { //k.logk
            string temp = strs[i];
            sort(begin(temp), end(temp));
            mp[temp].push_back(strs[i]);
        }
        for (auto it : mp) {
            result.push_back(it.second);
        }
        return result;
    }
};

// without sorting
// Tc = O(N*(k+26)) => O(N*K)

class Solution {
public:
    
    string generate(string &s) {
        int count[26] = {0};
        
        for(char &ch : s) {
            count[ch-'a']++;
        }
        
        string new_s;
        
        for(int i = 0; i<26; i++) {
            
            if(count[i] > 0) {
                new_s += string(count[i], i+'a');
            }
        }
        
        return new_s;
    }
    
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp;
        
        for(string &s : strs) {
            string new_s = generate(s);
            
            mp[new_s].push_back(s);
        }
        
        vector<vector<string>> result;
        for(auto &it : mp) {
            result.push_back(std::move(it.second));
        }
        
        return result;
        
    }
};

File: string/string-1/49.py
# Using Sorting
# T.C : O(n*klog(k))  (n = size of input, k = maximum length of a string in the input vector)
# S.C : O(n*k)

from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs):
        mp = defaultdict(list)
        for s in strs:
            temp = ''.join(sorted(s))
            mp[temp].append(s)
        return list(mp.values())
    
# without sorting
class Solution:
    def generate(self, s: str) -> str:
        count = [0] * 26
        
        for ch in s:
            count[ord(ch) - ord('a')] += 1
        
        new_s = []
        for i in range(26):
            if count[i] > 0:
                new_s.append(chr(i + ord('a')) * count[i])
        
        return ''.join(new_s)
    
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        mp = defaultdict(list)
        
        for s in strs:
            new_s = self.generate(s)
            mp[new_s].append(s)
        
        return list(mp.values())